### PR TITLE
Use build script to generate binding files

### PR DESCRIPTION
**Pull request by upsuper**
This is a WIP patch to make rust-bindgen a build time dependency of style component for stylo. This should make things more automatic. I convert majority of `regen.py` to `build_gecko.rs`.

It currently doesn't work on Windows, because of servo/rust-bindgen#271 (it works when I fix the generated file manually, though). I haven't tested other platforms.

It would break servo's CI for geckolib, because:
1. it needs libclang (which is probably easy to solve)
2. it needs `MOZ_OBJDIR` to be passed in so that it can generate bindings

@Manishearth @emilio Do you agree with this change? Do you have suggestion for the issues above?

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14244)

### CODE REVIEW

**Comment by Wafflespeanut** [comment]
While automating the bindings generation is a wonderful idea, I'm ambivalent about this for practical use, especially because of we'll be needing a gecko build for a simple `geckolib` build. We could still pass an argument to `./mach build-geckolib` for setting the `MOZ_OBJDIR`, but I feel this would be an unnecessary dependency while refactoring existing glue code or the FFI code, for example.

**Comment by Manishearth** [comment]
> While automating the bindings generation is a wonderful idea, I'm ambivalent about putting this for practical use, especially because of we'll be needing a gecko build for a simple geckolib build

Yeah. To me, build-time bindgen only makes sense when we have a merged repo. Until we have the double-CI workflow, geckolib _must_ build on its own so that we have some reasonable level of CI assurance.

(We can, of course, make build time bindgen an option that only runs when the MOZ_OBJDIR env var is present)

**Comment by Wafflespeanut** [comment]
> (We can, of course, make build time bindgen an option that only runs when the MOZ_OBJDIR env var is present. Folks wanting to update the in-tree structs file can run build time bindgen and copy the generated files)

Yes, we should consider this. We can make use of `mach` to help ease the workflow of the people who generate the bindings (@emilio, @Manishearth or @heycam) :smile:

**Comment by upsuper** [comment]
Making build time bindgen run only when MOZ_OBJDIR is present would add burden of maintaining sync between regen.py and build_gecko.rs. Also people would still need to check in binding files which they may forget when local have things done automatically.

Another option is that we drop this and continue using regen.py. Just let gecko's mach invoke that script explicitly. That way, we may need to figure out an approach to teach backends understand the dependency. That also means gecko need to vendor rust-bindgen manually.

**Comment by emilio** [comment]
This is great in my opinion, and it's probably where we want to go in the long term anyway, so +1 from me, though breaking the build in servo's CI is definitely a problem that we should probably fix.

**Comment by upsuper** [comment]
We may remove regen.py but still keep binding files in tree, and make build script use the in-tree files when `MOZ_OBJDIR` is not available. Then people can manually copy the binding files from target dir to in-tree location. Locating output dir may take some time, but I guess that wouldn't be more troublesome than manually running regen.py.

**Comment by Manishearth** [comment]
> Making build time bindgen run only when MOZ_OBJDIR is present would add burden of maintaining sync between regen.py and build_gecko.rs.

Not necessarily? I said we only run it when MOZ_OBJDIR is present, not that we should keep regen.py. Like I said, "Folks wanting to update the in-tree structs file can run build time bindgen and copy the generated files"

There's only a slight issue with this -- in my workflow when working on new things I usually sync and check the diff (since sometimes I mess up) and we no longer get to do that. Oh well.

**Comment by bholley** [comment]
@indygreg is currently working on [1] as a stopgap to autolander, and is doing something related to automatic binding generation. I'll let him weigh on on whether this makes things better or worse in the mean time.

[1] https://bugzilla.mozilla.org/show_bug.cgi?id=1317525

File: components/style/build.rs
Comment (Manishearth) [inline components/style/build.rs]: We need to also be able to trigger a rebuild if the contents of the objdir change, which is a bit harder to do here.

Perhaps make the build script unconditional and have it detect this?

File: components/style/gecko_bindings/bindings.rs
Comment (Manishearth) [inline components/style/gecko_bindings/bindings.rs]: please don't remove these. Commit the newly generated versions.

**Review by Manishearth — CHANGES_REQUESTED** [review]
See if you can make this optional.

Also, r?@emilio on the new bindings script? You're more familiar with the bindgen args. It overall looks correct to me, but I may have missed something.

**Comment by froydnj** [comment]
> Another option is that we drop this and continue using regen.py. Just let gecko's mach invoke that script explicitly. That way, we may need to figure out an approach to teach backends understand the dependency. That also means gecko need to vendor rust-bindgen manually.

Keeping `regen.py` requires a lot more plumbing for integration on the Gecko side; Gecko's build system strongly prefers that you generate files with Python scripts, not with Rust binaries.  Teaching it to build Rust binaries to invoke them ("natively" or via Python script wrapper) would be some work.  It would be preferable if all this stuff was invoked via Rust build scripts.

I'm not up to speed on the terminology (geckolib?), but it sounded to me like an automatic build-time solution + semi-manual generation via `mach build-geckolib` would work OK?  Please feel free to correct my ignorance. 😄

File: components/style/build.rs
Comment (upsuper) [inline components/style/build.rs]: That is done in build_gecko.rs which recursively adds header files which it may depend on to `cargo:rerun-if-changed`.

**Comment by Manishearth** [comment]
> I'm not up to speed on the terminology (geckolib?)

geckolib is the name of the stylo library built from the servo repo that gets linked into Firefox. It can currently be built independently of a Firefox build. Until the CI is all there, I'd like it to stay that way.

**Comment by upsuper** [comment]
Now the blockers are servo/rust-bindgen#272 and servo/rust-bindgen#274.

**Comment by upsuper** [comment]
It is almost done. Now build-time bindgen is optional. It would use the in-tree version if `MOZ_OBJDIR` is not available. There is still one problem that rust-bindgen has duplicate versions for package aster, which we need to resolve, otherwise test-tidy would complain.

**Comment by upsuper** [comment]
Travis CI's libclang version is too old... That is 3.4... rust-bindgen requires at least 3.8. Is there a chance we can upgrade the system we use there?

**Comment by upsuper** [comment]
The test-tidy failure is because of duplicate versions of aster. That is an issue of rust-bindgen which I have pr for: servo/rust-bindgen#275.

**Comment by nox** [comment]
Just saw this. Can we avoid the dependency on syntex_syntax?

**Comment by emilio** [comment]
You can look at how to install modern libclang on Travis on the bindgen CI script: https://github.com/servo/rust-bindgen/blob/master/ci/before_install.sh

**Comment by bholley** [comment]
@upsuper, how close are we on this? I think it's probably necessary to make our auto-incubator stuff work (currently the stylo builds are all failing from binding issues).

**Comment by upsuper** [comment]
@bholley it has almost completed. I already had a local patch which can use the in-tree bindings if the needed environment variable is not passed in. The remaining issue for this pr itself is to include libclang in the CI environment so that we don't break the CI. I can probably just take what rust-bindgen does.

I was waiting for [bug 1318981](https://bugzilla.mozilla.org/show_bug.cgi?id=1318981) because that's a prerequisite for this to work as expected. And I am currently waiting for @emilio's work to fix rust-bindgen, because I expect that would also change the original regen.py script quite a bit.

If the change to regen.py isn't coming soonish but this change is critical, I can try to complete this first, so that further changes would go to the new build script directly. I didn't realize that this has a very high priority.

**Comment by bholley** [comment]
Assuming this is going to be optional, we don't need to wait on Travis changes, right? I thought the idea was that the Travis/Homu tests would still used the checked-in bindings, and that build-time bindgen would be used on incubator.

**Comment by upsuper** [comment]
This is going to be optional at runtime of the build script, but to build the build script, we need libclang to present. clang-sys supports loading libclang at runtime, maybe we can make rust-bindgen support that as well. (But any change to rust-bindgen required would also mean we may be blocked on upgrading it.)

That also reminds me, does the Gecko CI have libclang right now?

**Comment by bholley** [comment]
> That also reminds me, does the Gecko CI have libclang right now?

Good question - @indygreg?

**Comment by upsuper** [comment]
> This is going to be optional at runtime of the build script, but to build the build script, we need libclang to present.

Oh, actually we can hide build-time bindgen behind a feature, and only enable that feature in stylo build, but not normal geckolib build, so that we don't need to depend on it in Servo's CI.

**Comment by froydnj** [comment]
> That also reminds me, does the Gecko CI have libclang right now?

The files exist in tooltool, but I am unsure of where to indicate they
should be downloaded, or what environment variables we need to setup for
cargo.

**Comment by upsuper** [comment]
> The files exist in tooltool, but I am unsure of where to indicate they should be downloaded, or what environment variables we need to setup for cargo.

It is 3.9 or 3.8?

We need it to be installed in some standard location, or its location is passed in via `LIBCLANG_PATH` environment.

clang-sys searches the following paths if no environment is provided on Linux:
> /usr/lib*
> /usr/lib*/*
> /usr/lib*/*/*
> /usr/local/lib*
> /usr/local/lib*/*
> /usr/local/lib*/*/*
> /usr/local/llvm*/lib

**Comment by upsuper** [comment]
I would say it is done now.

r? @emilio

**Review by emilio — APPROVED** [review]

**Pull request merged**
